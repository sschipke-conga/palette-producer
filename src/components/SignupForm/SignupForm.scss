.input-error {
  border: 1px solid red;
}

.form-label-input-div {
  height: 60px;
  margin-top: 20px;
  input {
    height: 100%;
    width: 100%;
  }
}

button {
  font-size: 24px;
  margin-top: 20px;
  height: 50px;
  width: 33%;
}