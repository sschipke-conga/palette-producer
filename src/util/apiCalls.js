export const createNewUser = async user => {
  let url = `${process.env.REACT_APP_PALETTE_PRODUCER_BACKEND_BASE_URL}/api/v1/signup`;
  let options = {
    method: "POST",
    body: JSON.stringify(user),
    headers: {
      "Content-Type": "application/json"
    }
  };
  let res = await fetch(url, options);

  if (res.status === 500) {
    throw Error("This username already taken");
  }

  if (!res.ok) {
    throw Error("Woops! Something went wrong");
  }

  return res.json();
};

export const loginUser = async user => {
  let url = `${process.env.REACT_APP_PALETTE_PRODUCER_BACKEND_BASE_URL}/api/v1/login`;
  let options = {
    method: "POST",
    body: JSON.stringify(user),
    headers: {
      "Content-Type": "application/json"
    }
  };
  let res = await fetch(url, options);

  if (res.status === 401) {
    throw Error("Username or password is incorrect");
  }

  if (!res.ok) {
    throw Error("Woops! Something went wrong");
  }

  return res.json();
};
