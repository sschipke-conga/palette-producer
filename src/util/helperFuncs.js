export const generateRandomHex = () => {
  let hexCode = "#" + Math.floor(Math.random() * 16777215).toString(16);
  if (hexCode.length < 7) {
    return generateRandomHex();
  }
  return hexCode;
};
