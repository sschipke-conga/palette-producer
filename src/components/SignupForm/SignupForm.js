import React, {Component} from 'react';
import {createNewUser} from '../../util/apiCalls';
import {Redirect} from 'react-router-dom';
import './SignupForm.scss'

class SignupForm extends Component {
  constructor() {
    super();
    this.state = {
      username: '',
      password: '',
      confirmPassword: '',
      isFormComplete: false,
      passwordError: false,
      hasError: false,
      error: ''
    };
  }

  handleChange = e => {
    this.setState({ [e.target.id]: e.target.value });
  };

  handleSubmit = async (e) => {
    e.preventDefault()
    const {password, confirmPassword, username} = this.state
    if (password === confirmPassword) {
      this.setState({passwordError:false})
      const newUser = {
        username,
        password
      }
      try {
        let res = await createNewUser(newUser)
        this.setState({username: '', password: '', confirmPassword: '', error:'', hasError: false,isFormComplete:true})
      localStorage.setItem("user", JSON.stringify({user_id: res.id, username: res.username}))
        this.props.loadProjects(res.id)
      } catch ({message}) { this.setState({hasError: true, error: message})}
    } else {
      this.setState({ error: 'Passwords do not match', passwordError: true})
    }
  };

  render() {
    const {error, passwordError, hasError, username, password, confirmPassword, isFormComplete} = this.state;
    if(isFormComplete) {
      return <Redirect to="/" />
    }
    const passwordErrorClass = passwordError ? "input-error" : "";
    const usernameErrorClass = hasError ? "input-error" : "";
    return (
      <form onSubmit={this.handleSubmit}>
        {hasError ? (<p className="error-message">{error}</p>):(<p className="form-message">Create an account to save your projects and the palettes you generate!</p>)}
        <div className="form-label-input-div">
          <label htmlFor="username">Username
            </label>
          <input
            type="text"
            className={usernameErrorClass}
            placeholder="Enter your username"
            id="username"
            required
            value={username}
            onChange={this.handleChange}
          />
        </div>
        <div className="form-label-input-div">
          {passwordError ? (<label className="error-message" htmlFor="password">{error}
          </label>) : (<label htmlFor="password"> Password
            </label>)}
          <input
            className={passwordErrorClass}
            type="password"
            placeholder="Enter your password"
            id="password"
            minLength="8"
            required
            value={password}
            onChange={this.handleChange}
          />
        </div>
        <div className="form-label-input-div">
          {passwordError ? (<label className="error-message" htmlFor="confirmPassword">{error}
          </label>) : (<label htmlFor="confirmPassword"> Confirm your password
            </label>)}
          <input
            className={passwordErrorClass}
            type="password"
            placeholder="Re-enter your password"
            id="confirmPassword"
            minLength="8"
            required
            value={confirmPassword}
            onChange={this.handleChange}
          />
        </div>
        <button type="submit">
          Sign Up!
        </button>
      </form>
    );
  }
}

export default SignupForm;