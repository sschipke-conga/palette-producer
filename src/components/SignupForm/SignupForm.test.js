import React from 'react';
import SignupForm from './SignupForm';
import { shallow } from 'enzyme';
import { createNewUser, loginUser} from '../../util/apiCalls';

jest.mock('../../util/apiCalls');

createNewUser.mockImplementation(() => {
  return Promise.resolve({
    ok: true,
    json: () => jest.fn().mockImplementation(() => {
      return Promise.resolve({
        id: 5,
        name: 'Steve',
        password: '1234'
      })
    })
  })
});

loginUser.mockImplementation(() => Promise.resolve({
  id: 5,
  name: 'Steve',
  password: '1234'
}));

describe('SignupForm', () => {
  let mockUser = {
    id: 5,
    name: 'Steve',
    email: 's@gmail.com',
    password: '1234'
  }
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<SignupForm
    />)

  })
  it('should match the initial snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  });
  describe('handleChange', () => {
    it('should update state if a user types in their username', () => {
      const mockEvent = { target: { id: 'username', value: 'Steve' } };
      expect(wrapper.state('username')).toEqual('');
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('username')).toEqual(mockEvent.target.value)
    })
    it('should update state if a user types in their password', () => {
      const mockEvent = { target: { id: 'password', value: 'password' } };
      wrapper.instance().setState({ username: 'Steve' });
      expect(wrapper.state('username')).toEqual('Steve')

      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('password')).toEqual(mockEvent.target.value);
      expect(wrapper.state('username')).toEqual('Steve');
    })
    it('should run handleChange when the inputs detect a change', () => {
      const mockNameEvent = { target: { id: 'username', value: 'Robbie' } };
      const mockPasswordEvent = { target: { id: 'password', value: 'password' } };
      wrapper.instance().handleChange = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.find('[id="username"]').simulate('change', mockNameEvent);
      wrapper.find('[id="password"]').simulate('change', mockPasswordEvent);
      expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockNameEvent);
      expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockPasswordEvent);
    })
  })
  describe('handleSubmit', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<SignupForm />)
    })
    let mockUser = {
      id: 4,
      username: "Susan",
      password: "password"
    }
    it('should prevent the default action when the form is submitted', () => {
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.find('form').simulate('submit', mockEvent);
      expect(mockEvent.preventDefault).toHaveBeenCalled()
    })
    it('should call createNewUser when called', () => {
      let mockNewUser = {
        username: "Susan",
        password: "password"
      }
      let mockEvent = {preventDefault: jest.fn()}
      createNewUser.mockImplementation(() => {
        return Promise.resolve(mockUser)
      })
      wrapper.instance().setState({ username: 'Susan', password: 'password', confirmPassword: 'password' });
      wrapper.instance().handleSubmit(mockEvent);
      expect(createNewUser).toHaveBeenCalledWith(mockNewUser);
    })
    it('should update with an error if the passwords don\'t match', () => {
      let mockEvent = { preventDefault: jest.fn() }
      createNewUser.mockImplementation(() => {
        return Promise.resolve(mockUser)
      })
      wrapper.instance().setState({ username: 'Susan', password: 'password', confirmPassword: 'wrong' });
      wrapper.instance().handleSubmit(mockEvent);
      expect(wrapper.state('passwordError')).toEqual(true);
      expect(wrapper.state('error')).toEqual('Passwords do not match')
    })
    it('should update its state to be logged in', async () => {
      let mockEvent = {preventDefault: jest.fn()}
      wrapper.instance().setState({ username: 'Susan', password: 'password', confirmPassword: 'password' });
      expect(wrapper.state('isFormComplete')).toEqual(false);
      await wrapper.instance().handleSubmit(mockEvent)
      expect(wrapper.state('isFormComplete')).toEqual(true);
    });
    it('should set an error message to state if something goes wrong', async () => {
      let mockEvent = {preventDefault: jest.fn()}
      createNewUser.mockImplementation(() => {
        throw Error('Woops')
      });
      wrapper.instance().forceUpdate()
      await wrapper.instance().handleSubmit(mockEvent);
      expect(wrapper.state('error')).toEqual('Woops')
    })
  })

  describe('alt snapShot', () => {
    it('should match the snapshot if a user is logged in', () => {
      let wrapper = shallow(<SignupForm />);
      wrapper.instance().setState({ isLoggedIn: true });
      expect(wrapper).toMatchSnapshot()
    })
  })
})

