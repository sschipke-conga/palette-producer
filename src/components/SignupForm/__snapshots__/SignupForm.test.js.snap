// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm alt snapShot should match the snapshot if a user is logged in 1`] = `
<form
  onSubmit={[Function]}
>
  <p
    className="form-message"
  >
    Create an account to save your projects and the palettes you generate!
  </p>
  <div
    className="form-label-input-div"
  >
    <label
      htmlFor="username"
    >
      Username
    </label>
    <input
      className=""
      id="username"
      onChange={[Function]}
      placeholder="Enter your username"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-label-input-div"
  >
    <label
      htmlFor="password"
    >
       Password
    </label>
    <input
      className=""
      id="password"
      minLength="8"
      onChange={[Function]}
      placeholder="Enter your password"
      required={true}
      type="password"
      value=""
    />
  </div>
  <div
    className="form-label-input-div"
  >
    <label
      htmlFor="confirmPassword"
    >
       Confirm your password
    </label>
    <input
      className=""
      id="confirmPassword"
      minLength="8"
      onChange={[Function]}
      placeholder="Re-enter your password"
      required={true}
      type="password"
      value=""
    />
  </div>
  <button
    type="submit"
  >
    Sign Up!
  </button>
</form>
`;

exports[`SignupForm should match the initial snapshot 1`] = `undefined`;
